### PR TITLE
UserApi Development Initialized

### PR DESCRIPTION
**Pull request by ajammil**
I did the first endpoint, but I need the feedback.
Also in the Mock test for the UserApi, I disabled the assertSent function as it gave me errors. Could you tell me what I did wrong?
Thank you in advance.

This PR is targeting issue #183 .

### CODE REVIEW

File: src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Map;
+
+import org.testng.annotations.Test;
+
+import com.cdancy.bitbucket.rest.BaseBitbucketMockTest;
+import com.cdancy.bitbucket.rest.BitbucketApi;
+//import com.cdancy.bitbucket.rest.BitbucketApiMetadata;
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: Lets remove this if it's not needed.

File: src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    null, null, null, null, null, 0, 2);
+            assertThat(up).isNotNull();
+            assertThat(up.errors()).isEmpty();
+            assertThat(up.size() == 1).isTrue();
+            assertThat(up.values().get(0).slug().equals("jcitizen")).isTrue();
+
+            final Map<String, ?> queryParams = ImmutableMap.of("filter", "jcitizen",
+                    "group","stash-users" , "permission", "ADMIN","start",0,"limit",2);
+            //assertSent(server, getMethod, restApiPath + BitbucketApiMetadata.API_VERSION
+            //        + "/users", queryParams);
```
Comment (cdancy) [inline src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: What is the failure you are seeing here?

File: src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    @GET
+    UserPage users(@Nullable @QueryParam("filter") String filter,
+                   @Nullable @QueryParam("group") String group,
+                   @Nullable @QueryParam("permission") String globalRole,
+                   @Nullable @QueryParam("permission.1") String projectRole,
+                   @Nullable @QueryParam("permission.1.projectKey") String projProjectKey,
+                   @Nullable @QueryParam("permission.1.projectId") String projProjectId,
+                   @Nullable @QueryParam("permission.2") String repoRole,
+                   @Nullable @QueryParam("permission.2.projectKey") String repoProjectKey,
+                   @Nullable @QueryParam("permission.3.projectId") String repoPrrojectId,
```
Comment (cdancy) [inline src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java]: So we could have N number of these which is going to make this a bit unruly and also limit folks if we only specify 3. We'll probably have to create an optional Binder which will take in a list of of these "permission objects" which we will then have to inject the values as query-params. Take a look [HERE](https://github.com/cdancy/bitbucket-rest/blob/master/src/main/java/com/cdancy/bitbucket/rest/binders/BindHookSettingsToPayload.java) and poke around the code base to see how we do this type of thing elsewhere. Let me know if you have any questions. Looks good so far.

File: src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java (hunk 1 above)
Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: alright, I will clean it up.

File: src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java (hunk 2 above)
Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: TravisCI is failing to pull, I want to show you the detailed errors from the build file. But I'm waiting for it to detect my commits.

Comment (cdancy) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: Do things fail locally when you run `./gradlew clean build mockTest`?

File: src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java (hunk 3 above)
Comment (ajammil) [reply src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java]: I'm sorry I misunderstood the documentation I thought permission was for global 1 was for project permissions and 2 for repo etc.. I didn't recognize they were the same thing but different filters.

I looked into what you told me, isn't that a way to bind some data to the payload? which is as much as I know should not be used with GET requests as their body is not processed? Please correct me if I misunderstood anything.

Comment (cdancy) [reply src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java]: Ignore the class name ... you can alter the request in general and append new query params/etc. Take a look at the below code where we do something like this:

https://github.com/cdancy/jenkins-rest/blob/master/src/main/java/com/cdancy/jenkins/rest/filters/ScrubNullFolderParam.java

File: src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java (hunk 2 above)
Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: Yes, they give me checkStyle error: Do not use Windows line endings

Comment (cdancy) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: Ok so not an issue with the test/assertion but more an issue with your gitconfig which you can play with to get this working as you'd expect.

Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: No, it doesn't even get to check assertions it fails from the beginning.

Comment (cdancy) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: Are you developing on windows or on a unix based machine? I do all development on unix so I'm not sure where that would be coming from. To fix things you _could_ just write the line out again :)

Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: I am using Ubuntu for my developments, I will try to figure out what's wrong and get back to you.

Comment (ajammil) [reply src/test/java/com/cdancy/bitbucket/rest/features/UserApiMockTest.java]: I found out where the Mock errors were, also I removed that error from my checkStyle so it's easier to build locally, I didn't stage it tho.

File: src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java (hunk 3 above)
Comment (ajammil) [reply src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java]: I am sorry I have to postpone it to the next Friday, as am having my midterm exams this week and I got a lot of studying to do.

Comment (cdancy) [reply src/main/java/com/cdancy/bitbucket/rest/features/UserApi.java]: @ajammil of course and take your time.

**Pull request closed without merging**